File: modules/jooby-apt/src/main/resources/io/jooby/internal/apt/Source.java
package ${packageName};
${imports}
@io.jooby.annotation.Generated(${className}.class)
public class ${generatedClassName} implements io.jooby.Extension {
    protected java.util.function.Function<io.jooby.Context, ${className}> factory;
${constructors}
    public ${generatedClassName}(${className} instance) {
       setup(ctx -> instance);
    }

    public ${generatedClassName}(io.jooby.SneakyThrows.Supplier<${className}> provider) {
       setup(ctx -> (${className}) provider.get());
    }

    public ${generatedClassName}(io.jooby.SneakyThrows.Function<Class<${className}>, ${className}> provider) {
       setup(ctx -> provider.apply(${className}.class));
    }

    private void setup(java.util.function.Function<io.jooby.Context, ${className}> factory) {
        this.factory = factory;
    }

${methods}
}

File: modules/jooby-apt/src/main/resources/io/jooby/internal/apt/Source.kt
package ${packageName}
${imports}
@io.jooby.annotation.Generated(${className}::class)
open class ${generatedClassName} : io.jooby.MvcExtension {
    private lateinit var factory: java.util.function.Function<io.jooby.Context, ${className}>

    ${constructors}
    constructor(instance: ${className}) { setup { instance } }

    constructor(provider: io.jooby.SneakyThrows.Supplier<${className}>) { setup { provider.get() } }

    constructor(provider: (Class<${className}>) -> ${className}) { setup { provider(${className}::class.java) } }

    constructor(provider:  io.jooby.SneakyThrows.Function<Class<${className}>, ${className}>) { setup { provider.apply(${className}::class.java) } }

    private fun setup(factory: java.util.function.Function<io.jooby.Context, ${className}>) {
      this.factory = factory
    }
${methods}
}
